Implement NonZero operator

atausz opened the pull request:
This pull request implements the NonZero operator for agate. See https://github.com/onnx/onnx/blob/master/docs/Operators.md#nonzero for the ONNX reference.

Note: We may need to transpose the result of this since pytorch computes the transposed version of this. However, as-is the current implementation matches the ONNX specification.

r? @thomas-stripe @oscar-stripe

johnynek commented on tests/src/test/scala/com/stripe/agate/tensor/TensorTest.scala:
```diff
@@ -824,4 +824,84 @@ object TensorTest extends Properties("TensorTest") {
+    val c1_1 = {
+      val t0 = tensor"[5, -1, 0, 4.4, 0, 666]"
+      val got = t0.nonZero
+      val expected = (tensor"[[0, 1, 3, 5]]").map(DataType.Int64)(_.toLong)
```
Instead of map and toLong does cast

https://github.com/stripe/agate/blob/f03e1b131e397c0c0e45a22db7cf0233c16703d2/core/src/main/scala/com/stripe/agate/tensor/Tensor.scala#L183

Work?

johnynek commented on tests/src/test/scala/com/stripe/agate/eval/OperationTest.scala:
```diff
@@ -252,4 +252,16 @@ object OperationTest extends Properties("OperationTest") {
+      test("NonZero", t0 :: Nil, List(expected), Map.empty)
+    }
+
+    c1
```
I don’t think we need to capture c1 and then return it. Can we just remove the val c1 and return test?

We used that pattern in other places when there were multiple claims. If we left some around with only 1 it is because e changed the code and got lazy about cleaning it up after we removed claims.

johnynek commented on core/src/main/scala/com/stripe/agate/tensor/Tensor.scala:
```diff
@@ -153,6 +155,38 @@ abstract class Tensor[D <: DataType] {
     }
   }
 
+  def nonZero: Tensor[DataType.Int64.type] = {
```
Can we add a scala doc comment of what this returns? I am trying to guess but even after reading the test examples I find it confusing. I guessed first it was 1 everywhere the original was nonZero. But that doesn’t seem right.

atausz commented:
@johnynek - I addressed your review comments. I agree that the method is a bit confusing - it actually returns the indices for which the tensor is non-zero (scala doc comment describes).

Note that also the final output of the operator now matches what PyTorch computes.

r? @johnynek

johnynek commented:
@atausz This is great. Let's merge.

Note, here are microsoft's test:

https://github.com/microsoft/onnxruntime/blob/master/onnxruntime/test/providers/cpu/tensor/nonzero_op_test.cc

which we could copy in if we find the transpose bit is not right, but we should discover that when training and scoring a model that doesn't match.
